### ISSUE TITLE
Separate DBs so that LabMan uses API to access Qiita DB

### ISSUE THREAD

**Issue by tanaes** [jdereus/labman#257]
It was realized that to do a soft rollout with parallel testing, it would be necessary with the current code to either 1) deploy on the live system (scary) or 2) have some sort of automated database updating from qiita-main to qiita-rc, so that new samples added to Qiita were made available to the wet lab via qiita-rc. 

Because LabMan is only accessing the study_sample table to get Sample IDs, Study IDs, and study titles, it was decided that for a similar amount of effort it would be possible to adapt LabMan to access these tables via the Qiita API and have a totally stand-alone database.

**Comment by josenavas** [comment]
Note that the initial design of the labman project (April 2017) was done using that approach. When the new design came into place (Summer 2017), it was identified that having separated databases was one of the fundamental mistakes of the initial design.

Note that the database does not only access those tables to get information, but the database is actually creating Foreign Keys to those tables. This creates strong consistency and side effects that were desired. The simplest example is that by just using the API, a user could delete an entire study from Qiita in which the wet lab has already plates the samples, loosing all the history about the study.

Note that the problem exposed in this issue is a problem of testing design, and the proposed solution is to break consistency on production environment to "alleviate" a testing pain. Of course, deploying on the live system is a bad idea, but there are no 2 databases, there is a single database (the Qiita one). The approach to be taken should be to test the system enough to have confidence that the tables added to the Qiita database and the changes performed by Labman on the data will not break Qiita.

To achieve this simply reproduce the steps performed on the main Qiita on the Qiita-rc and let's hammer the system. I highly recommend against breaking one of the fundamental design decisions on the system just because it makes it hard to test ATM. Note that once the initial test is done and labman is finally deployed for production, it should follow the same deployment structure as Qiita.

Are @jdereus and @AmandaBirmingham being consulted before these decisions/proposals are made? The project has a long story and arguments for all these quirks have been extensively made.

**Comment by tanaes** [comment]
This change proposal comes after a long discussion this morning with
@jdereus and @AmandaBirmingham. Apologies for the very brief explanation on
the issue; testing was not the *sole* reason for the proposed change... But
note that we had not discussed the particular foreign key issue. Jeff and
Amanda, does that change your opinion on this?

**Comment by tanaes** [comment]
...namely, that management of coordinated database testing/patching could be an issue going forward, along with issues associated with credentialing and access. @wasade, could you also chime in on your thoughts here?

**Comment by josenavas** [comment]
Thanks for providing a bit more background about the nature of this morning's discussion.

Some more background on some of those issues mentioned. The long term goal was to include Labman as part of Qiita. It was initially developed independently for 2 main reasons: initial development speed and being able to use python 3. Note that the plan is to eventually move Qiita to py3 (see issue https://github.com/biocore/qiita/issues/250).

After Labman is flagged to be production ready, Labman should follow the same deployment schedule as Qiita does, so that no coordinated database patching is needed (there is a single database, so all patches go to the same place).

I'm not sure about the issue with credentialing and access. Labman already limits which users have access to the system, so general Qiita user's should not have access to it. I know there was a discussion about the fact that for the users with access to Labman can indeed see all the data, but that issue can be  solved easier by adding user groups to the system.

Finally, regarding the FK. The issue is more conceptual: Labman actually changes the state of the data in Qiita: once in use on the labman pipeline, the data is "locked" in the sense that it can't be deleted. Unless there are alternative methods to lock the data and mechanisms in place in which Qiita can communicate to Labman regarding changes on the data, there will be data integrity and consistency issues. (Also note that Qiita initiated communication to Labman goes against the Qiita design).

I am happy to provide more information on these issues - as you know I spent a lot of time thinking/discussing/designing most of the system, and it went through more than 3 iterations.

**Comment by AmandaBirmingham** [comment]
From the discussion this morning, I was not under the impression that credentialing and access were an issue of having a shared database, but rather (a) separating the databases would have to be done with care to avoid breaking the access approach and (b) there was a driver for more nuanced access control (regardless of what database structure is used).  Item A @antgonza had a proposed handling for, and Item B was I believe tabled for the future.

Besides the testing situation, the concerns that I heard voiced about the shared database model focused on whether labman performance would impact qiita performance and vice versa (since labman has some db-intensive functionality), and also on whether testing time would become prohibitive (the idea as I understood it being that if there is one db then every time one changes anything about its structure one must retest everything that depends on that structure).  I'm listing these here not because I have an informed opinion on them (I don't) but as a record of the factors being considered.

@josenavas Could you say more about the locking approach used in labman to change the state of qiita data, and where in the codebase it is implemented?  This sounds like a key element of the discussion!

**Comment by antgonza** [comment]
Just to add to the discussion. 

The main idea (for Qiita dev) of the split is to add a new service via plugins. In other words, being able to have a full other service (currently labman, but you can imagine adding gnps, a new proteomics service, etc) via plugins. For example, in the current state, if we wanted to start a single instance of qiita and labman they would need to have 2 different workers, which fine. However, this gives us the opportunity to fully develop (with minimal development) the idea of a service plugin.

In the current state, the main changes are:
- user authentication: the plan is to add a handler that gets a user id and returns the hash (remember that only tools with valid credentials can query) so internally labman can do the verification 
- sample ids: add a handler to reply the sample info file, similar to the prep info file that currently exists. I can see the desirability of "strong consistencies" but I think not having them allows us to have multiple plugin services and each service is responsible of keeping track (something similar to what EBI-Qiita does).

**Comment by josenavas** [comment]
@AmandaBirmingham thanks for the explanation - there's no specific locking approach implemented in labman, it is using the default locking approach in the [Postgres DB](https://www.postgresql.org/docs/9.4/static/explicit-locking.html#LOCKING-TABLES). In short, when doing an update, delete or insert in a table, the command acquires a ROW EXCLUSIVE lock on the table being updated (plus an ACCESS SHARE on all the reference tables).

Regarding the db-intensive functionality: those operations are performed in tables that Qiita doesn't know about, and Labman doesn't directly change the tables that Qiita accesses (it has indirect effects, as explained earlier by the created references through FK). Based on this, those operations should not affect Qiita performance. However, if there is a serious concern, it will be nice to support that concern with benchmarks and data.

@antgonza thanks for adding more background information!

User authentication: I think that is fine. In fact, as discussed in previous meetings, it would be ideal to have the entire Qiita-DB functionality exported through a rest API and everything else consumes that rest API (including the Qiita-pet GUI web-based interface).

Sample ids: "Strong consistencies" was not a "nice to have" of Labman's requirements - it was a "must". In fact, it is probably the only requirement that actually justifies the creation of Labman. The wet lab already have procedures to perform all the steps outlined in the Labman workflow, but the main issue is that there is a disconnect between those procedures and the rest of the data (metadata, sequences, etc). The power of Labman actually comes from the fact that it is able to keep track of all this data and ensure a consistent view of all the data. If that is removed, then I don't think it is even worth spending time on the project...

Plugin services: I'm not sure about the motivation - GNPS, for example, is already supported with the current plugin system, so I don't see why others would not be the same. "Something similar to what EBI-Qiita does": in fact, that system works the other way around - Qiita is the client of EBI and Qiita IS responsible of ensuring that the data view between Qiita and EBI is consistent (right now Qiita is NOT automatically ensuring this, but rather somebody needs to tell Qiita to "MODIFY" the EBI data...). The main difference, is in this case Qiita is the client and EBI is the server.

Now, thinking on the bigger picture (and not necessarily only on Labman - which, again, it is supposed to be rolled into Qiita at some point). The idea is to make Qiita a server in which other services are clients. If some of those services require strong consistencies (as Labman does), the current approach will not support it, and I don't think the solution is to "relax" the requirements of other systems, but actually providing a system that supports such consistencies. For example, a possible solution would be to create some type of pub/sub mechanism in which the client services subscribe to a channel for the data that they care about. Qiita publishes the changes on those channels and then the services update their data as needed, based on the changes published by Qiita.

The pub/sub mechanism will not work for Labman. The reason is that the pub/sub mechanism assumes that Qiita is in total control of the data and it doesn't need to check with any other service to update the data. However this is not true with Labman. Labman has the power to prevent Qiita from deleting samples if those samples have already been plated. Again, this was part of the requirements and it was intended, not an undesired "side-effect". _Maybe_ some kind of locking mechanism could be implemented through a pub/sub and/or exposing some other rest API points, but I think the time to have a minimum working solution (including proper design, implementation and testing) is not in the usual KL timelines.

**Comment by antgonza** [comment]
@josenavas, I don't follow why "strong consistencies" are so important and perhaps they even will cause more problems. For example, at any point the sample information file (or even the prep info file) comes from Qiita to labman, in fact; AFAIK sample names are the only things used and come from Qiita to labman, not the other way around. Furthermore, and AFAIK, once a plate is created (with multiple samples) the relationship of samples/plates/etc ends there. Let's put it this way, in the worst scenario, if a sample is deleted in Qiita and the relationship is broken in labman, I think it's fine: due cause if this happens there are other problems (no permissions to process the sample, etc) and ignoring the sample is fine (but it will not change the fact that it was platted ...)

**Comment by josenavas** [comment]
The final results of the labman pipeline are (1) a Qiita compatible prep template and (2) the sample sheet for sequencing. If strong consistency is not followed, then the prep template will not be able to be added to Qiita. Also, the sample sheet for shotgun it's used to split the sequencing files in a per sample fashion and, eventually, upload those files to the Qiita studies as needed. With no such consistency, you're not solving any of the problems that labman was designed to solve.

In addition, note that I put an example in which a sample is deleted, but nothing prevents to the entire study to be deleted. In which case, Labman says that sequencing run X belongs to study Y in Qiita which... Oh! it doesn't exist! Let's track down the study owner, ask which is the new study id, add a mechanism to labman to update the study id, update all files generated, all sample ids everywhere, etc...

As a reminder, the first iteration of labman was implemented in a different database. This specific consistency problem was what triggered looong discussions over multiple weeks until it was found that having it in the same database was the best solution.

On of the main goals is to save hours and hours of Gail, Greg, Jeff, and others' time, which are usually the ones that have to figure out what is going on when there is the disconnect between the metadata and the wet lab procedures. Another main goal is to halt the sample processing as soon as possible if no metadata is available (this was strongly required by @rob-knight), which contradicts your previous comment:

>Let's put it this way, in the worst scenario, if a sample is deleted in Qiita and the relationship is broken in labman, I think it's fine: due cause if this happens there are other problems (no permissions to process the sample, etc) and ignoring the sample is fine (but it will not change the fact that it was platted ...)

**Comment by AmandaBirmingham** [comment]
(A) it is not accurate to say that strong consistency is the only feature that justifies Labman.  Although the wet lab has procedures to produce the required files, they are more manual than is desirable, and it is also desirable to track in a db all the internal steps to generate plates and pools.  Not saying strong consistency isn’t a very worthwhile aim but don’t undersell the rest of the effort! :)
(B) The early (I can’t say if they were the first) implementations of Labman database called for *reimplementation* of sample and study tables, which implied a great deal of synchronization pain.  Removing that duplication was definitely one of the reasons the move was made to share a database with Qiita.
(C) I agree with @josenavas that it is definitely NOT ok if a sample or study are deleted from Qiita after having been used in the wet lab.  If fact, if a critical aim is to ensure that once samples in Qiita are used in the wet lab, they can’t be screwed with in Qiita, then the current implementation doesn’t go *far enough*. Simple DB constraints will prevent deleting any record whose key is used as a foreign key elsewhere in the db—but they won’t keep the rest of the values for that record from being screwed up.  We may need a way in Qiita to explicitly mark samples/studies as “locked” (beyond just db locking during transactions) to prevent harm to the non-key info about samples.

**Comment by josenavas** [comment]
(A) Agree :-)
(C) That's the end goal. In fact, the plan was for Labman to not allow the processing of a sample if the metadata was not complete. However, since Qiita currently doesn't have a notion of marking the metadata complete (or satisfies minimum requirements), that functionality was pushed back until that was figured out. Note that this issue was not a lack of the Qiita mechanism, but on actually agreeing in a minimum set of metadata that could be automatically checked (not sure if there has been any advancements on this end). Thus, in the meantime Labman just requires the sample to be present in the Sample Information section (and hence locking it through the FK), but if these other restrictions are placed on the sample metadata for processing, extra mechanisms will need to be implemented.

**Comment by antgonza** [comment]
After testing different options offline, we decided that the most sane thing at this stage is to implement https://github.com/jdereus/labman/issues/270 vs. the solutions described here. Thus, closing this in favor of https://github.com/jdereus/labman/issues/270
